File: 2023/rust/src/day8/part1.rs
#![allow(dead_code)]

use std::collections::HashMap;

#[derive(Debug)]
enum Instruction {
    Left,
    Right,
}

#[derive(Debug)]
struct Instructions(Vec<Instruction>);

#[derive(Debug)]
struct Map(HashMap<String, (String, String)>);

fn parse_input(input: &str) -> (Instructions, Map) {
    let mut input = input.lines();

    let instructions = Instructions(
        input
            .next()
            .unwrap()
            .chars()
            .map(|c| match c {
                'L' => Instruction::Left,
                'R' => Instruction::Right,
                _ => panic!(),
            })
            .collect(),
    );

    input.next().ok_or("?").unwrap();
    let map = Map(input
        .map(|line| {
            let mut split = line.split(" = ");
            let key = split.next().ok_or("hold up").unwrap().to_string();
            let mut split = split.next().unwrap().split(", ");
            let left = split.next().unwrap().replace('(', "");
            let right = split.next().unwrap().replace(')', "");
            (key, (left, right))
        })
        .collect());

    (instructions, map)
}

fn part1(instructions: Instructions, map: Map) -> u32 {
    let mut position = &"AAA".to_string();

    for (steps, ins) in instructions.0.iter().cycle().enumerate() {
        if position == "ZZZ" {
            return steps as u32;
        }
        match ins {
            Instruction::Left => position = &map.0.get(position).unwrap().0,
            Instruction::Right => position = &map.0.get(position).unwrap().1,
        }
    }

    u32::MAX
}

#[cfg(test)]
mod test {
    use super::*;

    use test_case::case;

    #[ignore]
    #[case("ex1.txt")]
    #[case("ex2.txt")]
    #[case("input.txt")]
    fn test_input(input: &str) {
        let input = std::fs::read_to_string(format!("src/day8/{}", input)).unwrap();
        let (instructions, map) = parse_input(&input);
        dbg!(instructions, map);
    }

    #[case("ex1.txt" => 2)]
    #[case("ex2.txt" => 6)]
    #[case("input.txt" => 20777)]
    fn test_part1(input: &str) -> u32 {
        let input = std::fs::read_to_string(format!("src/day8/{}", input)).unwrap();
        let (instructions, map) = parse_input(&input);
        part1(instructions, map)
    }
}

File: 2023/8/day8_go/part1.go
package day8

import (
	"strings"
)

const (
	Left = iota
	Right
)

type Instructions []int

type Map map[string]Destinations

type Destinations struct {
	Left  string
	Right string
}

func ParseInput(input string) (instructions Instructions, mapIn Map) {
	mapIn = make(Map)

	for i, line := range strings.Split(input, "\n") {
		switch i {
		case 0:
			for _, char := range line {
				switch char {
				case 'L':
					instructions = append(instructions, Left)
				case 'R':
					instructions = append(instructions, Right)
				}
			}
		case 1:
			continue
		default:
			split := strings.Split(line, " = ")

			start := split[0]

			destinations := strings.Split(split[1], ", ")

			left := strings.TrimLeft(destinations[0], "(")
			left = strings.TrimRight(left, ")")

			right := strings.TrimLeft(destinations[1], "(")
			right = strings.TrimRight(right, ")")

			mapIn[start] = Destinations{left, right}
		}
	}

	return
}

func Part1(instructions Instructions, mapIn Map) (steps uint64) {
	position := "AAA"

	for i := 0; ; i = (i + 1) % len(instructions) {
		if position == "ZZZ" {
			return steps
		}
		switch instructions[i] {
		case Left:
			position = mapIn[position].Left
		case Right:
			position = mapIn[position].Right
		}
		steps++
	}

}
